### ISSUE TITLE
container/list should state that it is not thread-safe

### ISSUE THREAD

**Issue by RobMeades** [golang/go#25105]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
 go1.6.2

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
linux/amd64

### What did you do?
Accessed a `container/list` from two different `go func()` routines.

### What were you tripped-up by
A panic occurred in `list.PushBack()` due to a nil pointer reference.  This was because the other (timed) `go func()` routine was iterating through the list, processing entries, and then calling `list.Remove()` on the processed items.  It seems as though the threading in Golang is so good that the timed `go func()` is acting upon an item while it is in the middle of being added to the list with `list.PushBack()`.

### What would you like to happen as a result?
It would be a good idea to add to the top of the `container/list` documentation page that it is NOT thread-safe, just so that no-one else falls over this.  It took a few million attempts for me to hit it; an hour or two with loops iterating every 20 ms: it might be a very intermittent fatal bug for others, so worth a warning.

**Comment by randall77** [comment]
This is true of most of the standard library. Even built-in maps are not thread safe. I'm not sure we want to pepper every API with such a warning.
Did you run with the race detector?  It should have caught this bug immediately.

**Comment by bcmills** [comment]
OTOH, we certainly should mention the race detector in Effective Go. It seems we currently do not (#25107).

**Comment by RobMeades** [comment]
Fair enough: I do make use of channels etc. but in this case I'm synchronising a semi-synchronous stream with a synchronous one, filling in any gaps by re-inserting nearby data, and so a pair of lists with a timed function in the middle seems like the obvious answer.  Just need to remember that, despite its modernity, linked lists in Golang are as "exciting" as ever.
